### ISSUE TITLE
Web interface is very inaccessible to screen reader users

### ISSUE THREAD

**Issue by ndarilek** [concourse/concourse#328]
I'm a blind screen reader user attempting to use Concourse, but unfortunately your web interface is very inaccessible. It seems to rely heavily on graphical, icon-based links with no text content. Thus far I've downloaded the wrong command line client twice because I've had to download the binary and investigate it with `file` to determine if it was the right one. I'm also failing to start builds on a pipeline.

Fortunately there's a very easy fix. For any links that don't have text content (I.e. `<a ...>Text content</a>`) just add an `aria-label` attribute. So `<a ... aria-label="Text content"/>` will fix this. I can't submit a PR because I can't see the text content or infer the meanings of the graphical icons. :)

As an aside, once you've created and unpaused a pipeline, does the fly command line let you execute the first build? I'm following the tutorial git repo, have successfully created and unpaused the pipeline, but I don't immediately see how to start a build from the command line. The tutorial makes an abrupt shift from using the command line to popping into the web interface at this point, and that's where I hit all these unlabeled links. I'd normally pop into Slack and ask but Slack is also inaccessible. :/ Like what I see so far, but until the web interface sproutes some aria-labels it's pretty much unusable to me.

Thanks.

**Comment by vito** [comment]
Sorry for this, thanks for the heads-up. I'll prioritize this pretty high and see what we can get done before 1.0. Is the pipeline itself navigable from the main page? Curious if the SVG aspect causes problems or if it's just the buttons. A sizeable pipeline will have a lot of repeated text from the resources propagating through, which could get annoying.

The getting started guide doesn't mention it but there is a new `trigger-job` command now which you could use: http://concourse.ci/fly-trigger-job.html

**Comment by ndarilek** [comment]
Ah, tanks for the trigger-job pointer, I've been searching `fly -h` for 
something useful and not finding it. :) What version is this in? The 
Vagrant quickstart seems to pull down 0.74.0 which doesn't seem to 
include it. Where would I get a newer version of fly if I have an older 
Vagrant, and will the trigger-job command work with 0.74 (I.e. does it 
call an existing API that just wasn't previously integrated into the CLI?)

All I can do is click the pipeline name. From that point on it's pretty 
much unusable--I didn't even know there were SVGs, for instance.

I haven't worked with SVGs much myself, but here is an accessible SVG 
demo. Not sure if there's anything you can extrapolate from it: 
https://output.jsbin.com/yiwuru

Thanks for the quick response. Despite the headaches of the web UI, I 
like that the CLI and text editor are first-class citizens--very 
convenient for me. :)

**Comment by vito** [comment]
There was a delay in publishing the v0.75.0 Vagrant box due to a failure in our pipeline, but it should be out now. You may need to run `vagrant destroy`, then `vagrant box update`, and then `vagrant up` again to bring your VM up to date.

Also, to clarify: the entire pipeline view is one big SVG, with a list of pipelines in the sidebar just for higher-level navigation. The SVG shows the jobs and resources in the pipeline, visually arranged in fancy ways that don't really help you, so I was hoping you'd at least be able to navigate to the jobs in the pipeline. I'll check out your link - I'm concerned the whole thing may be a black box to screen readers right now.

We'll be working on a more simplified view of the pipeline eventually. The original use case is for CI displays and other things you want to see from far away, but it may end up being more useful in your case as well.

How easy would it be to get a proper screen reader set up on our end to test these things out?

**Comment by ndarilek** [comment]
Ah, I was using a different Vagrantfile that pulled in 0.74.0. Trying 
the quickstart method now.

Sorry, what I meant to say was that the only aspect of the pipeline that 
was accessible was its name (I.e. I could click on "helloworld" but 
everything else was unfortunately a black box.) If everything is more or 
less accessible via the command line then I'm OK for now but would 
definitely like to have access to the web UI at some point.

Screen reader setup is fairly easy, and all major platforms offer 
something for free. Probably the most spec compliant is NVDA for Windows 
(http://nvaccess.org). Under Linux there's Orca, which runs under GNOME 
and ships with most distros. OS X has VoiceOver built-in, enabled by 
cmd-F5 I believe. Your best bet is probably NVDA, since its parent 
foundation gets funded by Google/Mozilla to build against ARIA and other 
access technology specs. IOW, if you have something working well with 
NVDA then it'll probably work almost anywhere, and any remaining issues 
are likely with the screen readers themselves.

Thanks for being interested in making this happen.

**Comment by avtar** [comment]
@vito If you haven't already you could try using an accessibility evaluation tool such as [tota11y](http://khan.github.io/tota11y/). I did a quick check using the http://ci.concourse.ci/ hosted instance to confirm the issues @ndarilek reported. NVDA is also a great suggestion. Alternatively here are a couple of automated options:

* https://github.com/dequelabs/axe-core/
* https://github.com/GoogleChrome/accessibility-developer-tools/ (used by tota11y)

**Comment by waharnum** [comment]
I've done some work in the past on SVG accessibility and found the following helpful, although it's a few years old: 

- https://www.paciellogroup.com/blog/2013/12/using-aria-enhance-svg-accessibility/
- http://www.sitepoint.com/tips-accessible-svg/

In my experience using **role** was pretty key to making sure the intent of an SVG was properly communicated to a screen reader.
